### PR TITLE
FSDP Tests

### PR DESCRIPTION
**Pull request by bcui19**
Adds basic tests for FSDP. The tests are similar to that of DDP, Deepspeed. We should still add checkpointing tests for FSDP (@abhi-mosaic is drafting the tests)

# What issue(s) does this change relate to?
CO-1263

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/utils/misc.py [hunk 1]
```diff
@@ -28,6 +30,21 @@ def is_model_ddp(model: torch.nn.Module) -> bool:
     return isinstance(model, DistributedDataParallel)
 
 
+def is_model_fsdp(model: torch.nn.Module) -> bool:
+    """Whether ``model`` is an instance of a :class:`.FullyShardedDataParallel`."""
+    try:
+        from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+        is_fsdp = False
+        # This code basically looks to see if the model
+        # is wrapped with FSDP
```
Comment (mvpatel2000) [inline composer/utils/misc.py:39]:
```suggestion
        # Check if model is wrapped with FSDP
```

**Review by mvpatel2000 — COMMENTED** [review]
LGTM! If this becomes blocking, I can approve, though I'd want @abhi-mosaic to take a look today before we cut it since he's more familiar w FSDP.

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -113,11 +113,11 @@ def _filter_metrics(metrics: Dict[str, Metric], metric_names: Optional[List[str]
...
 
 
-def _validate_precision(precision: Precision, device: Device, deepspeed_enabled: bool):
+def _validate_precision(precision: Precision, device: Device, deepspeed_enabled: bool, fsdp_enabled: bool):
     if isinstance(device, DeviceCPU) and precision != Precision.FP32:
         raise ValueError(f'{precision} is not supproted for CPU training.')
-    if not deepspeed_enabled and precision == Precision.FP16:
-        raise ValueError('FP16 precision is only supported when training with DeepSpeed.')
+    if not (deepspeed_enabled or fsdp_enabled) and precision == Precision.FP16:
+        raise ValueError('FP16 precision is only supported when training with DeepSpeed or FSDP.')
```
Comment (abhi-mosaic) [inline composer/trainer/trainer.py:120]: This is incorrect, `FP16` is only a DeepSpeed-allowed precision type

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -413,6 +413,40 @@ def test_deepspeed(
...
+    ):
+
+        fsdp_config = {
+            'sharding_strategy': 'FULL_SHARD',
+            'min_params': 1e8,
+            'cpu_offload': False,
+            'mixed_precision': 'DEFAULT',
+            'backward_prefetch': 'BACKWARD_PRE',
+            'activation_checkpointing': False,
+            'activation_ocpu_offload': False,
```
Comment (abhi-mosaic) [inline tests/trainer/test_trainer.py:433]:
```suggestion
            'activation_cpu_offload': False,
```

File: composer/trainer/trainer.py (hunk 2 above)
Comment (abhi-mosaic) [reply composer/trainer/trainer.py:120]: And it should probably be removed in a future refactor..

File: tests/trainer/test_ddp.py [hunk 4]
```diff
@@ -159,6 +160,19 @@ def test_ddp(device: str, world_size: int, deepspeed: bool, tmp_path: pathlib.Pa
...
+    fsdp_config = None
+    if fsdp:
+        fsdp_config = {
+            'sharding_strategy': 'FULL_SHARD',
+            'min_params': 1e8,
+            'cpu_offload': False,
+            'mixed_precision': 'DEFAULT',
+            'backward_prefetch': 'BACKWARD_PRE',
+            'activation_checkpointing': False,
+            'activation_ocpu_offload': False,
```
Comment (abhi-mosaic) [inline tests/trainer/test_ddp.py:172]:
```suggestion
            'activation_cpu_offload': False,
```

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -468,10 +502,11 @@ def test_precision(
         should_error = False
         ctx = contextlib.nullcontext()
         if device == 'cpu' and precision != Precision.FP32:
-            ctx = pytest.raises(ValueError, match='not supproted for CPU training')
+            ctx = pytest.raises(ValueError, match='not supproted for CPU training.')
             should_error = True
         elif precision == Precision.FP16:
-            ctx = pytest.raises(ValueError, match='FP16 precision is only supported when training with DeepSpeed')
+            ctx = pytest.raises(ValueError,
+                                match='FP16 precision is only supported when training with DeepSpeed or FSDP.')
```
Comment (abhi-mosaic) [inline tests/trainer/test_trainer.py:509]: Same as above, `FP16` is only valid for DeepSpeed.

**Review by abhi-mosaic — APPROVED** [review]
LGTM, thank you!

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -468,10 +512,10 @@ def test_precision(
         should_error = False
         ctx = contextlib.nullcontext()
         if device == 'cpu' and precision != Precision.FP32:
-            ctx = pytest.raises(ValueError, match='not supproted for CPU training')
+            ctx = pytest.raises(ValueError, match='not supproted for CPU training.')
```
Comment (dakinggg) [inline tests/trainer/test_trainer.py:515]: "supproted" -> "supported"

**Pull request merged**